File: app/src/main/java/jp/bellware/echo/filter/FadeOut.java
package jp.bellware.echo.filter;

/**
 * フェードアウト
 */
public class FadeOut {

    /**
     * 全体の長さ
     */
    private int length;

    /**
     * フェードアウト区間
     */
    private int duration;

    /**
     * 現在処理位置
     */
    private int index;

    public FadeOut(int length, int duration) {
        this.length = length;
        this.duration = duration;
    }

    public float filter(float s){
        float v;
        if(index > length - duration){
            v = s*(length - index)/duration;
        }else{
            v = s;
        }
        index += 1;
        return v;
    }
}

File: app/src/main/java/jp/bellware/echo/filter/FadeOut.kt
package jp.bellware.echo.filter

/**
 * フェードアウト
 * @param length 全体の長さ
 * @param duration フェードアウト期間
 */
class FadeOut(
        private val length: Int,
        private val duration: Int) {

    /**
     * 現在処理位置
     */
    private var index: Int = 0

    /**
     * フィルターをかける
     */
    fun filter(s: Float): Float {
        val v: Float
        if (index > length - duration) {
            v = s * (length - index) / duration
        } else {
            v = s
        }
        index += 1
        return v
    }
}
